common/new-bet.ts

```diff
@@ -30,38 +40,185 @@ export type BetInfo = {
+      : outcome === 'YES'
+      ? Math.min(matchedBet.limitProb, limitProb ?? 1)
+      : Math.max(matchedBet.limitProb, limitProb ?? 0)
+
+    const poolAmount =
+      limit === undefined
+        ? amount
+        : Math.min(amount, calculateCpmmAmount(cpmmState, limit, outcome))
+
+    const { shares, newPool, newP, fees } = calculateCpmmPurchase(
+      cpmmState,
+      poolAmount,
+      outcome
+    )
+    const newState = { pool: newPool, p: newP }
+
+    return {
+      maker: {
+        matchedBetId: null,
+        shares,
+        amount: poolAmount,
+        state: newState,
+        fees,
+        timestamp,
+      },
+      taker: {
+        matchedBetId: null,
+        shares,
+        amount: poolAmount,
+        timestamp,
+      },
+    }
+  }
+
+  // Fill from matchedBet.
+  const matchRemaining = matchedBet.orderAmount - matchedBet.amount
+  const shares = Math.min(
+    amount /
+      (outcome === 'YES' ? matchedBet.limitProb : 1 - matchedBet.limitProb),
+    matchRemaining /
+      (outcome === 'YES' ? 1 - matchedBet.limitProb : matchedBet.limitProb)
+  )
+
+  const maker = {
+    bet: matchedBet,
+    matchedBetId: 'taker',
+    amount:
+      shares *
+      (outcome === 'YES' ? 1 - matchedBet.limitProb : matchedBet.limitProb),
+    shares,
+    timestamp,
+  }
+  const taker = {
+    matchedBetId: matchedBet.id,
+    amount:
+      shares *
+      (outcome === 'YES' ? matchedBet.limitProb : 1 - matchedBet.limitProb),
+    shares,
+    timestamp,
+  }
+  return { maker, taker }
+}
+
+export const getBinaryCpmmBetInfo = (
+  outcome: 'YES' | 'NO',
+  betAmount: number,
   contract: CPMMBinaryContract | PseudoNumericContract,
-  loanAmount: number
+  limitProb: number | undefined,
+  unfilledBets: LimitBet[]
 ) => {
-  const { shares, newPool, newP, fees } = calculateCpmmPurchase(
-    contract,
-    amount,
-    outcome
+  const sortedBets = sortBy(
+    unfilledBets.filter((bet) => bet.outcome !== outcome),
+    (bet) => (outcome === 'YES' ? bet.limitProb : -bet.limitProb),
+    (bet) => bet.createdTime
   )
 
-  const { pool, p, totalLiquidity } = contract
-  const probBefore = getCpmmProbability(pool, p)
-  const probAfter = getCpmmProbability(newPool, newP)
+  const takers: fill[] = []
+  const makers: {
+    bet: LimitBet
+    amount: number
+    shares: number
+    timestamp: number
+  }[] = []
+
+  let amount = betAmount
+  let cpmmState = { pool: contract.pool, p: contract.p }
+  let totalFees = noFees
+
+  let i = 0
+  while (true) {
+    const matchedBet: LimitBet | undefined = sortedBets[i]
+    const fill = computeFill(amount, outcome, limitProb, cpmmState, matchedBet)
+    if (!fill) break
+
+    const { taker, maker } = fill
+
+    amount -= taker.amount
+
+    if (maker.matchedBetId === null) {
+      // Matched against pool.
+      cpmmState = maker.state
+      totalFees = addObjects(totalFees, maker.fees)
+      takers.push(taker)
+    } else {
+      // Matched against bet.
+      takers.push(taker)
```

Comment:
there should also be a creator fee assessed on the taker which should be added to totalFees

Reply:
I guess you can leave that for v 2.

Reply:
I guess I'll leave it for v2. I actually was planning to do it before.